<a href="https://colab.research.google.com/github/Grecia329/EcuacionesDiferencialesParciales-/blob/main/Pr%C3%A1ctica_1(J).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Práctica 1 (J) Juegos con Cartas

# Juego 1

In [15]:
import numpy as np
import random

In [16]:
cartas = np.arange(52)
print(cartas)

[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23
 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47
 48 49 50 51]


Obtener un número entero entre 0 y 51, y guardarlo dentro de una variable carta.

In [17]:
carta = int(np.random.randint(0,52))

print(carta)

15


**Simular 1000 partidas**

In [18]:
n = 1000000

juego = 0


for i in range(n):
  carta = int(np.random.randint(0,52))
  if carta < 26:
    juego += 1
  else:
    juego -= 1

# print(juego)

**Calcular la ganancia promedio:**

$$
\text{promedio} = \frac{\text{ganancia total}}{\text{# de juegos}}
$$

In [19]:
promedio = juego / n

print(promedio)

0.00198


$\therefore$  La simulación muestra que la ganancia promedio tiende a 0 conforme aumenta el número de repeticiones, por lo que el Juego 1 es un juego justo.

# Juego 2

In [ ]:
import numpy as np

cartas = np.arange(52)
print(cartas)

Cada 13 números vuelve a empezar el mismo valor
pero cambia el palo automáticamente

In [20]:
(cartas % 13) + 1

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13,  1,  2,  3,  4,
        5,  6,  7,  8,  9, 10, 11, 12, 13,  1,  2,  3,  4,  5,  6,  7,  8,
        9, 10, 11, 12, 13,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12,
       13])

Se calcula el valor esperado REAL matemático, es decir, se están revisando todas las combinaciones posibles.

La probabilidad de cada par es:
$$
p = (\frac{1}{52})(\frac{1}{51})
$$

In [21]:
Ej = 0

valores = (cartas % 13) + 1
p = (1/52)*(1/51)
for ci in valores:
  for cj in valores:
    if ci != cj:
      if ci+cj >= 15:
        Ej += p
      else:
        Ej -= p

print(Ej)

-0.07239819004524868


Se va jugar un millón de veces.

In [22]:
n = 10**6

suma = 0

for i in range(n):
  c1 = int(random.random()*52) + 1  # carta uno en [1,52] Cada ejecución te dará valores diferentes, todos entre 1 y 52. (pone el +1, para que el rango vaya de 1 a 52, y no de 0 a 52)
  v1 = (c1 % 13) + 1                # valor en [1,13] un número entero aleatorio entre 1 y 13
  mv = int(random.random()*51) + 1  # maximo valor en [1,51] que se sumará a c1
  c2 = ( (c1 +  mv) % 52 ) + 1      # carta dos entre [1,52] sin repetir c1
  v2 = (c2 % 13) + 1                # valor en [1,13]
  if v1 + v2 >= 15:
    suma += 1
  else:
    suma -= 1

print("La probabilidad estimada es: ", suma/n)

La probabilidad estimada es:  -0.077164


Es decir, en promedio pierdes ≈ 0.077 pesos por cada jugada.
Por ejemplo, al jugar 1000 veces pierdes 77 pesos.

# Conclusión

El Juego 1 es un juego justo ya que la ganancia esperada tiende a cero.
El Juego 2 presenta una esperanza negativa, por lo que el jugador pierde dinero en promedio.
Por lo tanto, conviene participar en el Juego 1.